ported/omb/rdm_latency.c

```diff
@@ -0,0 +1,422 @@
+	if (ret) {
+		FT_PRINTERR("fi_cq_open", ret);
+		goto err1;
+	}
+
+	/* Open completion queue for recv completions */
+	ret = fi_cq_open(dom, &cq_attr, &rcq, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_cq_open", ret);
+		goto err2;
+	}
+
+	memset(&av_attr, 0, sizeof(av_attr));
+	av_attr.type = fi->domain_attr->av_type ?
+			fi->domain_attr->av_type : FI_AV_TABLE;
+	av_attr.count = 2;
+	av_attr.name = NULL;
+
+	/* Open address vector (AV) for mapping address */
+	ret = fi_av_open(dom, &av_attr, &av, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_av_open", ret);
+		 goto err3;
+	 }
+
+	return 0;
+
+err3:
+	fi_close(&rcq->fid);
+err2:
+	fi_close(&scq->fid);
+err1:
+	return ret;
+}
+
+static int bind_ep_res(void)
+{
+	int ret;
+
+	/* Bind Send CQ with endpoint to collect send completions */
+	ret = fi_ep_bind(ep, &scq->fid, FI_SEND);
+	if (ret) {
+		FT_PRINTERR("fi_ep_bind", ret);
+		return ret;
+	}
+
+	/* Bind Recv CQ with endpoint to collect recv completions */
+	ret = fi_ep_bind(ep, &rcq->fid, FI_RECV);
+	if (ret) {
+		FT_PRINTERR("fi_ep_bind", ret);
+		return ret;
+	}
+
+	/* Bind AV with the endpoint to map addresses */
+	ret = fi_ep_bind(ep, &av->fid, 0);
+	if (ret) {
+		FT_PRINTERR("fi_ep_bind", ret);
+		return ret;
+	}
+
+	ret = fi_enable(ep);
+	if (ret) {
+		FT_PRINTERR("fi_enable", ret);
+		return ret;
+	 }
+
+	return ret;
+}
+
+static int init_fabric(void)
+{
+	int ret;
+	uint64_t flags = 0;
+
+	/* Get fabric info */
+	ret = fi_getinfo(FT_FIVERSION, NULL, NULL, flags, hints, &fi);
+	if (ret) {
+		FT_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
+
+	/* Open fabric */
+	ret = fi_fabric(fi->fabric_attr, &fab, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_fabric", ret);
+		goto err1;
+	}
+
+	/* Open domain */
+	ret = fi_domain(fab, fi, &dom, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_domain", ret);
+		goto err2;
+	}
+
+	/* Open endpoint */
+	ret = fi_endpoint(dom, fi, &ep, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_endpoint", ret);
+		goto err3;
+	}
+
+	/* Allocate endpoint resources */
+	ret = alloc_ep_res();
+	if (ret)
+		goto err4;
+
+	/* Bind EQs and AVs with endpoint */
+	ret = bind_ep_res();
+	if (ret)
+		goto err5;
+
+	return 0;
+
+err5:
+	free_ep_res();
+err4:
+	fi_close(&ep->fid);
+err3:
+	fi_close(&dom->fid);
+err2:
+	fi_close(&fab->fid);
+err1:
+	return ret;
+}
+
+static int init_av(void)
+{
+	void *addr;
+	size_t addrlen = 0;
+	int ret;
+
+	fi_getname(&ep->fid, NULL, &addrlen);
+	addr = malloc(addrlen);
+	assert(addr);
+	ret = fi_getname(&ep->fid, addr, &addrlen);
+	if (ret != 0) {
+		FT_PRINTERR("fi_getname", ret);
+		return ret;
+	}
+
+	addrs = malloc(numprocs * addrlen);
+	assert(addrs);
+
+	FT_Allgather(addr, addrlen, addrs);
+
+	fi_addrs = malloc(numprocs * sizeof(fi_addr_t));
+	assert(fi_addrs);
+
+	/* Insert address to the AV and get the fabric address back */
+	ret = fi_av_insert(av, addrs, numprocs, fi_addrs, 0, &fi_ctx_av);
+	if (ret != numprocs) {
+		FT_PRINTERR("fi_av_insert", ret);
+		return ret;
+	}
+
+	free(addr);
+
+	return 0;
+}
```

Comment:
The above functions look like a duplicate of what was added further up.  Please move into a common area.